core/src/main/scala/org/apache/spark/rdd/BlockRDD.scala

```diff
@@ -66,7 +67,9 @@ class BlockRDD[T: ClassTag](@transient sc: SparkContext, @transient val blockIds
     blockIds.foreach { blockId =>
       sc.env.blockManager.master.removeBlock(blockId)
     }
-    _isValid = false
+    if (_setInvalid) {
```

Comment:
Also, I will add unit tests to test this behavior for BlockRDD and WALBackedBlockRDD.